public/js/test/mochitest/head.js

```diff
@@ -0,0 +1,176 @@
+    while (count < eventRepeat) {
+      yield _afterDispatchDone(dbg.store, type);
+      count++;
+      info(type + " dispatched " + count + " time(s)");
+    }
+  });
+}
+
+function waitForThreadEvents(dbg, eventName) {
+  info("Waiting for thread event '" + eventName + "' to fire.");
+  const thread = dbg.toolbox.threadClient;
+
+  return new Promise(function(resolve, reject) {
+    thread.addListener(eventName, function onEvent(eventName, ...args) {
+      info("Thread event '" + eventName + "' fired.");
+      thread.removeListener(eventName, onEvent);
+      resolve.apply(resolve, args);
+    });
+  });
+}
+
+function waitForState(dbg, predicate) {
+  return new Promise(resolve => {
+    const unsubscribe = dbg.store.subscribe(() => {
+      if (predicate(dbg.store.getState())) {
+        unsubscribe();
+        resolve();
+      }
+    });
+  });
+}
+
+const waitForPaused = Task.async(function* (dbg) {
+  // We want to make sure that we get both a real paused event and
+  // that the state is fully populated. The client may do some more
+  // work (call other client methods) before populating the state.
+  return Promise.all([
+    yield waitForThreadEvents(dbg, "paused"),
+    yield waitForState(dbg, state => dbg.selectors.getPause(state))
+  ]);
+});
+
+const initDebugger = Task.async(function* (url, ...sources) {
+  const toolbox = yield openNewTabAndToolbox(url, "jsdebugger");
+  const win = toolbox.getPanel("jsdebugger").panelWin;
+  const store = win.Debugger.store;
+  const { getSources } = win.Debugger.selectors;
+
+  const dbg = {
+    actions: win.Debugger.actions,
+    selectors: win.Debugger.selectors,
+    getState: store.getState,
+    store: store,
+    toolbox: toolbox,
+    win: win
+  };
+
+  yield Promise.all(sources.map(url => {
+    return waitForState(dbg, state => {
+      return getSources(state).some(s => s.get("url").includes(url));
+    });
+  }));
+
+  return dbg;
+});
+
+function findSource(dbg, url) {
+  const sources = dbg.selectors.getSources(dbg.getState());
+  const source = sources.find(s => s.get("url").includes(url));
+
+  if(!source) {
+    throw new Error("Unable to find source: " + matchingStr);
+  }
+
+  return source.toJS();
+}
+
+function selectSource(dbg, url) {
+  info("Selecting source: " + url);
+  const source = findSource(dbg, url);
+  dbg.actions.selectSource(source.id);
+
+  return waitForDispatch(dbg, "LOAD_SOURCE_TEXT");
+}
+
+function stepOver(dbg) {
+  info("Stepping over");
+  dbg.actions.stepOver();
+  return waitForPaused(dbg);
+}
+
+function stepIn(dbg) {
+  info("Stepping in");
+  dbg.actions.stepIn();
+  return waitForPaused(dbg);
+}
+
+function stepOut(dbg) {
+  info("Stepping out");
+  dbg.actions.stepOut();
+  return waitForPaused(dbg);
+}
+
+function resume(dbg) {
+  info("Resuming");
+  dbg.actions.resume();
+  return waitForThreadEvents(dbg, "resumed");
+}
+
+function addBreakpoint(dbg, sourceId, line, col) {
+  return dbg.actions.addBreakpoint({ sourceId, line, col });
```

Comment:
I think it is good to keep the API similar to what the components call, that way we exercise the same behavior. We can't guarantee it's the same, but it's probably good to try.

The `addBreakpoint` command I setup assumes it's the selected source and passes along the `getTextForLine` param for that reason. 

```js
function addBreakpoint(line, { source } = {}) {
  source = source || selectors.getSelectedSource();
  return actions.addBreakpoint({
    sourceId: source.get("id"),
    line
  }, {
    getTextForLine: l => debuggerWindow.cm.getLine(l - 1).trim()
  });
}
```

If we want to pass along a source that is not the selected source, I'd propose that this action first selects that source and then adds the breakpoint. This will keep our tests behaving the same way our app does and keep us from getting into weird states.

```js
function addBreakpoint(line, { source } = {}) {
  if (source.get("url") != selectors.getSelectedSource().get("url")) {
    selectSource(source);
  }

  source = source || selectors.getSelectedSource();
  return actions.addBreakpoint({
    sourceId: source.get("id"),
    line
  }, {
    getTextForLine: l => debuggerWindow.cm.getLine(l - 1).trim()
  });
}
```

One anti-goal is pausing in a test and the debugger looks broken. An example of this is the breakpoint list not having text and not seeing the breakpoint which was just added.